### ISSUE TITLE
GPU delegate silently miscomputes a broadcast ADD on a BATCH_MATMUL result — every masked attention is wrong

### ISSUE THREAD

**Issue by john-rocky** [google-ai-edge/LiteRT#8593]
## Summary

On the ML Drift GPU delegate, `softmax(q @ kᵀ * scale + mask[1, 1, S, S])` returns
plausible-looking but wrong probabilities. That expression is ordinary masked
attention — what `transformers`, `diffusers` and most hand-written attention emit
for a causal or padding mask.

There is no error, no NaN, no partition fallback. The delegate reports **zero
unsupported operations** (`Replacing 7 out of 7 node(s) with delegate (LITERT_CL)
node, yielding 1 partitions`), and **the same `.tflite` on the same device's CPU
accelerator is exact** (`corr 1.000000`, `max|diff| 5.96e-08`).

This is a correctness bug, not a performance note. Minimal repro below: five
1–3 KB fp32 graphs, runnable in about a minute.

## What is actually wrong

Only the **logits** are wrong. Everything a sanity check would look at still holds
on the failing graph:

- the softmax rows still **sum to 1** (measured range `[0.999885, 1.000132]`),
- the **causal mask is still honoured** — measured probability mass above the
  diagonal is exactly `0.000e+00`,
- **no NaN, no inf** — the output is finite and bounded in `[0, 1]`,
- **row 0 is bit-exact (`max|diff| = 0.000e+00`); every later row is wrong
  (`max|diff| = 6.6e-01`).** Row 0 attends only to key 0, so it is the one row
  insensitive to key mixing. A broken RoPE has the identical signature, because
  RoPE is the identity at position 0 — that is the trap, and it cost us a probe
  round.

In the FLUX.2-klein 4B text encoder where we first hit this, the padding mask was
likewise still honoured, and the run reported success end to end — while the
generated image came out at correlation **0.016** against the reference.

## Minimal repro

Five fp32 graphs (`S=64, H=8, D=32`), converted with
`litert_torch.convert(module, args)`:

```python
class BmmBroadcastAdd(nn.Module):          # WRONG on GPU
    def forward(self, query, key, mask):   # mask: [1, 1, S, S]
        logits = torch.matmul(query, key.transpose(-1, -2)) * SCALE
        return torch.softmax(logits + mask, dim=-1)

class BmmFullAdd(nn.Module):               # exact on GPU
    def forward(self, query, key, mask):   # mask: [1, H, S, S]
        logits = torch.matmul(query, key.transpose(-1, -2)) * SCALE
        return torch.softmax(logits + mask, dim=-1)

class InputBroadcastAdd(nn.Module):        # exact on GPU — discriminator #1
    def forward(self, logits, mask):       # logits is a graph INPUT, mask: [1, 1, S, S]
        return torch.softmax(logits + mask, dim=-1)

class BmmBroadcastAddTapped(nn.Module):    # exact on GPU — discriminator #2
    def forward(self, query, key, mask):   # identical arithmetic to BmmBroadcastAdd
        scaled = torch.matmul(query, key.transpose(-1, -2)) * SCALE
        return scaled, torch.softmax(scaled + mask, dim=-1)   # matmul also an output
```

Three facts pin the trigger:

1. `BmmBroadcastAdd` and `BmmFullAdd` lower to the **identical op histogram** —
   `RESHAPE ×2, TRANSPOSE, BATCH_MATMUL, MUL, ADD, SOFTMAX`. The only difference
   is whether the `ADD`'s right operand is implicitly broadcast across the head
   axis. The broadcast one is wrong.
2. `InputBroadcastAdd` applies that same broadcast `ADD` to a graph **input**
   rather than to a `BATCH_MATMUL` result, and is exact. **So a broadcast `ADD`
   on its own is fine.**
3. `BmmBroadcastAddTapped` is `BmmBroadcastAdd` with the matmul result *also*
   returned as an output — same arithmetic, one extra output — and its softmax is
   exact. **So the miscompute is in how the pair is fused, not in either op.**

Runnable repro (export script, fixed inputs, host references, device runner,
comparison script): https://gist.github.com/john-rocky/dbd0b5da65c6af76157e3a444a4bb224

```bash
python build_repro.py --check      # exports the graphs; CPU corr 1.000000 each
./run_on_device.sh bmm_bcast_add   # GPU: MISCOMPUTED, corr 0.61
./run_on_device.sh bmm_full_add    # GPU: matches host, corr 1.000000
CPU=1 ./run_on_device.sh bmm_bcast_add   # same device, CPU: corr 1.000000
```

## Measured, on the graphs above

Pixel 8a, `CompiledModel`, LiteRT 2.1.3. Every graph fully delegates in one
partition — `LITERT_CL` on GPU, `TfLiteXNNPackDelegate` on CPU.

| graph | GPU | CPU (same device) |
|---|---|---|
| `bmm_bcast_add` — `softmax(q @ kᵀ * scale + mask[1,1,S,S])` | **MISCOMPUTED, corr 0.612677** | corr 1.000000 |
| `bmm_full_add` — `softmax(q @ kᵀ * scale + mask[1,H,S,S])` | corr 1.000000 | — |
| `bmm_no_mask` — `softmax(q @ kᵀ * scale)` | corr 0.999998 | — |
| `input_bcast_add` — `softmax(logits + mask[1,1,S,S])` | corr 1.000000 | — |
| `bmm_bcast_add_tapped` — matmul also an output | corr 1.000000 | — |

**It is not a precision issue.** With `SetPrecision(kFp32)` (`accepted=YES`) the
failing graph returns `corr 0.612705` — unchanged. Under fp32 the row sums are
`1.000000` to six decimals, so the softmax is perfectly normalized over wrong
logits.

## Additional probes

Run during isolation at the FLUX.2-klein encoder's shapes (`S, D, H = 512, 128,
32`), with the encoder's real mask:

| probe | GPU |
|---|---|
| `softmax(q @ kᵀ * scale + mask[1,1,S,S])` | corr 0.18 |
| `softmax(q @ kᵀ + mask[1,1,S,S])` — no scale | corr 0.08 |
| `softmax((q * scale) @ kᵀ + mask[1,1,S,S])` — scale folded into q | corr 0.18 |
| a `reshape` between the `ADD` and the `SOFTMAX` | corr 0.18 → not a softmax-fusion issue |
| a broadcast `MUL`/`ADD` on a plain input tensor, axis 1 and axis 2 | exact |
| a 32-way `CONCATENATION`, and a 4D `BATCH_MATMUL`, in isolation | exact |
| **`softmax(q @ kᵀ * scale + mask[1,H,S,S])`** | **exact** |
| `softmax(q @ kᵀ * scale) * keep; p /= p.sum(-1)` | exact (algebraic fallback) |

## Workaround

Pass the mask pre-expanded to `[1, num_heads, S, S]`. Zero extra ops, identical
formula, bit-exact. It costs one 33 MB fp32 input at `S=512, H=32`, so it is a
workaround, not a fix.

## Impact

Any encoder-style or causal transformer running on the GPU delegate with the
standard `[1, 1, S, S]` mask is exposed, and it fails silently. "Op-compat PASS"
and "0 unsupported ops" are not correctness signals for this class of bug.

## Environment

- Device: Pixel 8a — Google Tensor G3, `GLES: ARM, Mali-G715, OpenGL ES 3.2
  v1.r54p2`, Android 16
- Runtime: LiteRT **2.1.3** (`libLiteRt.so` and `libLiteRtClGlAccelerator.so`
  taken from `litert-2.1.3.aar`; SHA-1 verified identical)
- API: `CompiledModel`; GPU = ML Drift OpenCL (`LITERT_CL`, `delegate_opencl.cc`),
  CPU control = XNNPACK, same binary, same device
- Converter: `litert-torch`
- Graphs: fp32, so quantization is not in the suspect list

**Comment by john-rocky** [comment]
## Fixed in 2.1.6

Re-ran the attached repro against the official Maven AARs on the same Pixel 8a, holding the graphs fixed and swapping only the runtime `.so`:

| probe | 2.1.3 | 2.1.5 | 2.1.6 |
|---|---|---|---|
| `bmm_bcast_add` (the repro) | corr 0.612677 | 0.612677 | **1.000000** |
| masked attention at H=24, S=256 | corr 0.498593 | — | **0.999999** |
| masked attention at H=8, S=512 | corr 0.524628 | — | **0.999998** |

All at 7/7 `LITERT_CL`. 2.1.3 and 2.1.5 are bit-identical to each other, so the fix landed in the 2.1.6 ML Drift.

The 2.1.6 run needs a harness rebuilt against the v2.1.6 `litert_cc_sdk.zip`, since `libLiteRt.so` dropped `LiteRtSetEnvironmentOptionsValue` and `LiteRtGetCompiledModelEnvironment`. To rule the rebuilt harness out as the variable: the control graphs `bmm_no_mask` (max |diff| 2.545e-04 / 1.134e-03) and `bmm_full_add` (1.309e-03) return digit-identical fp16 values on both harnesses, and `bmm_bcast_add`'s 2.1.6 output now matches `bmm_full_add` exactly. Only the graph containing the `BATCH_MATMUL → implicitly-broadcast ADD` pair changed.

## Scoping the trigger — my original report over-generalised it

Probing each broadcast axis of a 4D `BATCH_MATMUL` result separately on **2.1.3**:

| shape | broadcast axis | 2.1.3 |
|---|---|---|
| `bmm[1,H,S,S] + bias[1,1,S,S]` | head (axis 1) | corr 0.628486 — **the bug** |
| `bmm[1,H,S,S] + bias[1,H,1,S]` (ALiBi) | query (axis 2) | 1.000000 |
| `bmm[1,H,S,S] + bias[1,H,S,1]` | key (axis 3) | 1.000000 |
| `bmm[W,H,S,S] + bias[1,H,S,S]` (windowed attention) | batch (axis 0) | 0.999999 |
| `bmm[1,H,S,S] * per_head[1,H,1,1]` | axes 2,3 | 1.000000 |

So only a **size-1 axis-1 (head) broadcast against a produced rank-4 tensor** was ever affected — not broadcasting in general. ALiBi bias, window relative-position bias and per-head score scaling were already correct.

The same fix also closes a sibling I hit separately: a dim-1 broadcast `MUL`, `[1,1,C,T] × [1,K,C,T]`, which returned source 0's product for every `k`. It goes corr **0.900981 → 1.000000** on 2.1.6 at 12/12 `LITERT_CL`. Same axis, same shape family — one fix, two bugs.

## Request

Nothing in the [2.1.6 release notes](https://github.com/google-ai-edge/LiteRT/releases/tag/v2.1.6) mentions this. Could it be added? Anyone still on ≤2.1.5 is silently shipping wrong masked attention — full delegation, softmax rows still summing to 1, mask structure still honoured — with no signal that a fix exists.

Happy to close this once someone can confirm the fix was intended and identify the commit.

**Comment by smilingday** [comment]
Thanks for reporting and happy to hear it's fixed in 2.1.6.

**Comment by smilingday** [comment]
@Linchenn could you check the issue, and whether ML drift release in 2.16 has a right fix for this? If so, any LiteRT release notes suggestion for this?

**Comment by smilingday** [comment]
This is fixed and we have updated release notes, thanks.
